Add melodic.

clalancette opened the pull request:
This adds melodic to the buildfarm configuration.  A few things to note:

1.  This is branched off of, and should be merged back into, the production branch.
1.  I ended up rationalizing the names of files in ways that made sense to me, though these files differ from what was in Lunar (but those differ from what was in Kinetic, so I'm not sure if there is a set standard).  Let me know if this is a problem.
1.  I haven't yet created the ros-buildfarm-melodic mailing list, but I will soon.
1.  I haven't tested this.  Is there any way to test before merging/deploying?

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Add melodic.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

dirk-thomas commented on index.yaml:
```diff
@@ -57,6 +57,21 @@ distributions:
+      ds: melodic/release-stretch-build.yaml
+      dsv8: melodic/release-stretch-arm64-build.yaml
+      uxhf: melodic/release-armhf-build.yaml
+      uxv8: melodic/release-arm64-build.yaml
```
These two keys should not be `ux` (for Ubuntu Xenial) but `ub` (for Ubuntu Bionic). The yaml file name included Xenial in Lunar to clarify that the build file only targets a single Ubuntu distro.

clalancette replied:
Ah, thanks.  Fixed in b737bac

dirk-thomas commented on melodic/doc-build.yaml:
```diff
@@ -0,0 +1,58 @@
+---
+canonical_base_url: http://docs.ros.org
+documentation_type: rosdoc_lite
+jenkins_job_priority: 89
```
The priorities shouldn't be equal but give preference to newer distros. It might be necessary to shuffle priorities to make room for the new ones.

Same below for the other build files.

clalancette replied:
I've shuffled around the priorities in 192e632 and bc96a4a.  I think I've got everything in the same relative priority order, with melodic inserted as appropriate.  A look at the output of `list_priorities.py` from someone else would be helpful, though.

dirk-thomas replied:
The Jenkins master is currently provisioned to only support priorities from 1 to 99.

clalancette replied:
> The Jenkins master is currently provisioned to only support priorities from 1 to 99.

Ah, that stinks :(.  OK, will collapse the numbers down.

dirk-thomas replied:
Maybe we need to make more room. Or change some of the priorities of the maintenance jobs. We should just make sure to have enough room for the next several releases to not run into the problem next year again.

clalancette replied:
There is actually quite a bit of the range we are not using (our lowest is 58, but even if we are leaving room for other things, we can go way lower than that).  I'll move the low end much further down, and insert some space between groupings so we have some breathing room.

dirk-thomas replied:
The maintenance jobs are using priorities in the lower range. And they need to have a higher prio then the package specific jobs. So you can't just use smaller numbers. See https://github.com/ros-infrastructure/ros_buildfarm/search?utf8=%E2%9C%93&q=priority%3D&type=

mikaelarguedas replied:
What is the policy for EOL distros? Do we need to keep the EOL jobs priorities in orde ? Or can we just reassign them? this way if we never have more than 4 distros we don't have to modify the range

dirk-thomas replied:
That would require to shift the priorities of all existing distros which is a significant reconfiguration effort I would rather avoid. We should just make sure that we have a large enough range for future releases (e.g. have "room" for the next ten year - by then the infrastructure will likely have been changed anyway).

clalancette replied:
That's true.  At the very least, I can collapse the "jade" priorities into the "indigo" numbers, for instance; since we never have jade jobs, that shouldn't really be a problem.

(I think we should probably just remove jade completely, but that is a different PR)

clalancette replied:
> We should just make sure that we have a large enough range for future releases (e.g. have "room" for the next ten year - by then the infrastructure will likely have been changed anyway).

The problem is that we have 6 "classes" of jobs: PR priority, commit priority, source priority, basic binary priority, doc priority, and extended binary priority.  If we want to leave room for 6 more years (say), we would need at least 60 spots.  But because of the pressure on the lower end from maintenance jobs, I think we only have 50 (the highest maintenance priority I see in a simple search is 40; I'm leaving the whole 40 range empty just for safety).

The nicest thing would be to allow the buildfarm to have more priority numbers.  How hard is it to change it?

dirk-thomas replied:
We can change that easily in the settings but that should be reflected in the provisioning too. Changing priorities of existing jobs once is ok (we might not get around it). But that change should be future proof.

clalancette replied:
> We can change that easily in the settings but that should be reflected in the provisioning too. Changing priorities of existing jobs once is ok (we might not get around it). But that change should be future proof.

Makes sense.  I think we have to change them now to fit into ranges, so I'll go ahead and do that with the understanding that we'll need to change the buildfarm before we deploy this.

dirk-thomas commented on melodic/release-arm64-build.yaml:
```diff
@@ -0,0 +1,59 @@
+  emails:
+  - ros-buildfarm-melodic@googlegroups.com
+  - clalancette+buildfarm@osrfoundation.org
+  maintainers: false
```
Why is this disabled?

Same below.

clalancette replied:
Mistake on my part.  Fixed in 755223d and f2d637d.

clalancette committed:
Switch ux -> ub.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Change up the jenkins priorities.

Melodic is always first, followed by Lunar, Kinetic, and
Indigo.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Fix up the job priorities to put melodic first.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Reenable arm64 build.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Make sure to mail the maintainers.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Rearrange the priorities, leaving room for newer distributions.

This depends on a reconfigure of the ROS Jenkins buildfarm to allow
priorities up to at least 109.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Add in documentation on how priorities are handled.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

dirk-thomas commented on PRIORITY_NOTES.md:
```diff
@@ -0,0 +1,17 @@
+| Source                       | 70 - 79       |
+| Basic Binary (amd64)         | 80 - 89       |
+| Documentation                | 90 - 99       |
+| Extended Architecture Binary | 100 - 109     |
```
Since this only seems to need a range 10 numbers bigger I would suggest to shift all numbers by 10 down instead and modify the very few hard coded priorities of maintenance jobs to make room in the are from 40-49.

This will avoid requiring changes to the deployment. Worst case some pull request jobs in several years have the same prio as some maintenance jobs.

clalancette replied:
So, I ended up moving back into the 1-99 range, and moved down some of the maintenance jobs in this PR: https://github.com/ros-infrastructure/ros_buildfarm/pull/500 .  At least we know we won't have any collisions for 5 years.

clalancette committed:
Update to fit into the 1-99 range currently configured on the buildfarm.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

dirk-thomas commented on PRIORITY_NOTES.md:
```diff
@@ -0,0 +1,17 @@
+This file contains some notes on the Jenkins job priorities that are configured in this repository.
+
+The ROS buildfarm is currently setup with a Jenkins plugin that manages job priorities.  The priorities are inverse numerically based; that is, a lower number means a higher priority.  As of this writing, the buildfarm is configured to have a range of 1 to 109.  There are periodic maintenance jobs that use priorities 1 to 49, thus those priorities should never be used for distribution jobs.
```
Please start each sentence on a new line.

clalancette replied:
Sure, will fix that up.

dirk-thomas reviewed (approved):
Beside the last comment this looks good to me.

When merging this all existing jobs need to be reconfigured to use the modified priorities. That will take a lot of time to run.

mikaelarguedas commented on PRIORITY_NOTES.md:
```diff
@@ -0,0 +1,17 @@
+| Pull Request                 | 40 - 49       |
+| Commit                       | 50 - 59       |
+| Source                       | 60 - 69       |
+| Basic Binary (amd64)         | 70 - 79       |
```
Nit: this applies to any non-ARM job at the moment (x86 and amd64)

clalancette replied:
> Nit: this applies to any non-ARM job at the moment (x86 and amd64)

Didn't we drop support for x86 a while back, though?

dirk-thomas replied:
In Lunar, yes. But Indigo and Kinetic are still building x86.

clalancette replied:
Ah, got it.  Thanks for the clarification.

clalancette committed:
Cleanup the PRIORITY_NOTES.md file.

In particular, put each sentence on its own line, and fix up
the numbers to reflect reality a bit.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>
